src/Controllers/PacksController.php

```diff
@@ -0,0 +1,536 @@
+                ->values();
+
+        $lessonContent =
+            $parentLessons->where('id', $lessonId)
+                ->first();
+
+        $nextChild = $parentLessons->getMatchOffset($lessonContent, 1);
+        $previousChild = $parentLessons->getMatchOffset($lessonContent, -1);
+
+        $thisLesson['next_content_type'] = 'lesson';
+
+        if (!$pack) {
+            $pack =
+                $this->contentService->getByChildIdWhereParentTypeIn($thisPackBundle['id'], ['pack'])
+                    ->first();
+        }
+        $pack = $this->isOwnedOrLocked($pack, auth()->id());
+
+        $thisLesson['is_owned'] = $pack['is_owned'];
+
+        $incompleteLesson =
+            $thisLesson['related_lessons']->where('completed', '!=', true)
+                ->where('id', '!=', $lessonId)
+                ->first();
+
+        if (empty($incompleteLesson)) {
+            $incompleteBundles =
+                $this->contentService->getByParentId($pack['id'])
+                    ->where('completed', '!=', true);
+
+            $thisLesson['next_content_type'] = 'bundle';
+            $thisLesson['bundle_count'] = $pack->fetch('bundle_count');
+            $bundleSiblings = $this->contentHierarchyService->getByParentIds([$pack['id']]);
+
+            foreach ($bundleSiblings as $index => $bundleSibling) {
+                if ($bundleSibling['child_id'] == $thisPackBundle['id']) {
+                    $nextBundleId =
+                        (array_key_exists($index + 1, $bundleSiblings)) ? $bundleSiblings[$index + 1]['child_id'] :
+                            null;
+
+                    if (!$nextBundleId) {
+                        $nextBundleId = $incompleteBundles->first()['child_id'];
+                    }
+
+                    $incompleteLesson = $this->contentService->getById($nextBundleId);
+                    if ($incompleteLesson) {
+                        $incompleteLesson['lesson_count'] = $this->contentHierarchyService->countParentsChildren(
+                            [$incompleteLesson['id']]
+                        )[$incompleteLesson['id']];
+                    }
+                }
+            }
+        }
+
+        $thisLesson['parent'] = $thisPackBundle;
+        $thisLesson['parent']['current_lesson'] = $incompleteLesson;
+
+        $thisLesson['next_lesson'] = $nextChild;
+        $thisLesson['previous_lesson'] = $previousChild;
+
+        $this->vimeoVideoDecorator->decorate(new Collection([$thisLesson]));
+
+        return ResponseService::content($thisLesson);
+    }
+
+    /**
+     * @param Request $request
+     * @param $id
+     * @return JsonResponse
+     * @throws NonUniqueResultException
+     */
+    public function jumpToNextLesson(Request $request, $id)
+    {
+        $pack = $this->contentService->getById($id);
+
+        $thisPackBundle = null;
+
+        if ($pack['type'] == 'pack') {
+
+            $lessonsProgressRows =
+                $this->contentRepository->connection()
```

Comment:
I think this query may be refactored into a repository or service.

Reply:
Agree. Moved the query into a service. Thanks!